tarodb/assets_common.go

```diff
@@ -308,29 +312,81 @@ type FetchGenesisStore interface {
 func fetchGenesis(ctx context.Context, q FetchGenesisStore,
 	assetID int32) (asset.Genesis, error) {
 
-	// Now we fetch the genesis information that so far we
-	// only have the ID for in the address record.
+	// Now we fetch the genesis information that so far we only have the ID
+	// for in the address record.
 	gen, err := q.FetchGenesisByID(ctx, assetID)
 	if err != nil {
 		return asset.Genesis{}, fmt.Errorf("unable to fetch genesis: "+
 			"%w", err)
 	}
 
-	// Next, we'll populate the asset genesis information which
-	// includes the genesis prev out, and the other information
-	// needed to derive an asset ID.
+	// Next, we'll populate the asset genesis information which includes
+	// the genesis prev out, and the other information needed to derive an
+	// asset ID.
 	var genesisPrevOut wire.OutPoint
 	err = readOutPoint(bytes.NewReader(gen.PrevOut), 0, 0, &genesisPrevOut)
 	if err != nil {
 		return asset.Genesis{}, fmt.Errorf("unable to read outpoint: "+
 			"%w", err)
 	}
 
+	var metaHash [32]byte
+	copy(metaHash[:], gen.MetaDataHash)
+
 	return asset.Genesis{
 		FirstPrevOut: genesisPrevOut,
 		Tag:          gen.AssetTag,
-		Metadata:     gen.MetaData,
+		MetaHash:     metaHash,
 		OutputIndex:  uint32(gen.OutputIndex),
 		Type:         asset.Type(gen.AssetType),
 	}, nil
 }
+
+// maybeUpsertAssetMeta inserts a meta on disk and returns the primary key of
+// that meta if metaReveal is non nil. Otherwise, this returns a NULL int32.
+func maybeUpsertAssetMeta(ctx context.Context, db UpsertAssetStore,
+	assetGen *asset.Genesis, metaReveal *proof.MetaReveal) (sql.NullInt32, error) {
+
+	var (
+		assetMetaID sql.NullInt32
+
+		metaHash [32]byte
+		metaBlob []byte
+		metaType sql.NullInt16
+
+		err error
+	)
+
+	switch {
+	// If there's a meta reveal with this asset genesis, then we'll also be
+	// inserting a blob and meta type.
+	case metaReveal != nil:
+		metaHash = metaReveal.MetaHash()
+		metaBlob = metaReveal.Data
+		metaType = sql.NullInt16{
+			Int16: int16(metaReveal.Type),
+			Valid: true,
+		}
+
+	// Otherwise, we'll just be inserting only the meta hash. At a later
+	// time, the reveal/blob can also be inserted.
+	case assetGen != nil:
+		metaHash = assetGen.MetaHash
+
+	// In the default case, there's no actual meta reveal, so we'll just
+	// use the meta hash of all zeroes.
```

Comment:
Ah, I guess that answers my previous question about the zero hash. So we can make the `meta_data_id` non-null in the DB, which removes one instance of `NullInt32`.

Reply:
So with the current logic, we'll end up inserting all zeroes for the hash+type, then return that primary key for `UpsertAssetMeta`. Those that have no meta then end up pointing towards that.